aspnetcore/signalr/hub-filters.md

```diff
@@ -0,0 +1,193 @@
+        try
+        {
+            return await next(invocationContext);
+        }
+        catch (Exception ex)
+        {
+            Console.WriteLine($"Exception calling '{invocationContext.HubMethodName}'");
+            throw ex;
+        }
+    }
+
+    // Optional method
+    public Task OnConnectedAsync(HubLifetimeContext context, Func<HubLifetimeContext, Task> next)
+    {
+        return next(context);
+    }
+
+    // Optional method
+    public Task OnDisconnectedAsync(HubLifetimeContext context, Exception exception, Func<HubLifetimeContext, Exception, Task> next)
+    {
+        return next(context, exception);
+    }
+}
+```
+
+Filters are very similar to middleware, `next` will invoke the next filter and the final filter will invoke the hub method. Filters can also store the result from awaiting `next` and run logic after the hub method has been called before returning the result from `next`.
+
+If a filter wants to skip a hub method invocation we recommend throwing a `HubException` and not calling `next` so the client will receive an error if it was expecting a result.
+
+## Using hub filters
+
+When writing the filter logic, try to make it generic by using attributes on hub methods instead of checking for hub method names.
+
+For example, lets write a filter that will check an argument to a hub method for banned phrases and replace any phrases it finds with "***".
+For the purposes of this example we'll assume a `LanguageFilterAttribute` class is defined and it has a property named `FilterArgument` that can be set when using the attribute.
+
+First, place the attribute on the hub method that has a string argument that we want to clean up.
+```csharp
+public class ChatHub
+{
+    [LanguageFilter(filterArgument: 0)]
+    public async Task SendMessage(string message, string username)
+    {
+        await Clients.All.SendAsync("SendMessage", $"{username} says: {message}");
+    }
+}
+```
+
+Next, define a hub filter that will check for the attribute and replace banned phrases in an argument to the hub method with "***".
+```csharp
+public class LanguageFilter : IHubFilter
+{
+    // populated from a file or inline
+    private List<string> bannedPhrases = new List<string>();
+
+    public async ValueTask<object> InvokeMethodAsync(HubInvocationContext invocationContext, Func<HubInvocationContext, ValueTask<object>> next)
+    {
+        var languageFilter = (LanguageFilterAttribute)Attribute.GetCustomAttribute(invocationContext.HubMethod, typeof(LanguageFilterAttribute));
+        if (invocationContext.HubMethodArguments.Count > languageFilter.FilterArgument &&
+            invocationContext.HubMethodArguments[languageFilter.FilterArgument] is string str)
+        {
+            foreach (var bannedPhrase in bannedPhrases)
+            {
+                str = str.Replace(bannedPhrase, "***");
+            }
+            if (invocationContext.HubMethodArguments is object[] arguments)
```

Comment:
This is making me want to change HubMethodArguments back to `object[]`.

Reply:
People should just create a new array and change it 😄 . But maybe we can just voliate all the rules and make it an object array

Reply:
We should either switch HubMethodArguments back to `object[]` *or* only include the else branch in the sample. Making HubMethodArguments an `IReadOnlyList<object>` only to cast it to `object[]` in the docs is sending mixed messages.